### ISSUE TITLE
CT executes post_* hooks in the wrong order

### ISSUE THREAD

**Issue by jcpetruzza** [erlang/otp#7397]
**Describe the bug**
When a suite runs with two or more hooks, the order in which the post_xxx_per_yyy callbacks are called should be the inverse of the one for the respective pre_xxx_per_yyy callbacks, but is currently the same.

**To Reproduce**
Let's use two suites `cth_a` and `cth_b` that just log each time a callback is executed:

```[erlang]
%% file:cth_a.erl
-module(cth_a).
-include("cth_gen.inc").
```
```[erlang]
%% file:cth_b.erl
-module(cth_b).
-include("cth_gen.inc").
```

```[erlang]
-module(foo_SUITE).

-include_lib("stdlib/include/assert.hrl").

-export([suite/0,  all/0,  groups/0, test_foo/1]).

suite() -> [{ct_hooks, [{cth_a, [], 0}, {cth_b, [], 1}]}].
all() ->  [{group, my_group}].
groups() -> [{my_group, [], [test_foo]}].
test_foo(_Config) ->  ?assert(false).
```

```[erlang]
-export([
    init/2,
    pre_init_per_suite/3, post_init_per_suite/4,
    pre_end_per_suite/3, post_end_per_suite/4,
    pre_init_per_group/4, post_init_per_group/5,
    pre_end_per_group/4, post_end_per_group/5,
    pre_init_per_testcase/4, post_init_per_testcase/5,
    pre_end_per_testcase/4, post_end_per_testcase/5
]).

init(_Id, _Opts) ->
  {ok, undefined}.

-define(LOGGING_CALL(X),
    ct:print("~p -- ~p", [?MODULE, ?FUNCTION_NAME]),
    ct:sleep({seconds, 1}),
    X
).

%%% SUITE
pre_init_per_suite(_Suite, InitData, State) ->
    ?LOGGING_CALL({InitData, State}).

post_init_per_suite(_Suite, Config, _Return, State) ->
    ?LOGGING_CALL({Config, State}).

pre_end_per_suite(_Suite, EndData, State) ->
    ?LOGGING_CALL({EndData, State}).

post_end_per_suite(_Suite, _Config, _Return, State) ->
    ?LOGGING_CALL({ok, State}).

%%% GROUP
pre_init_per_group(_Suite, _Group, InitData, State) ->
    ?LOGGING_CALL({InitData, State}).

post_init_per_group(_Suite, _Group, Config, _Return, State) ->
    ?LOGGING_CALL({Config, State}).

pre_end_per_group(_Suite, _Group, EndData, State) ->
    ?LOGGING_CALL({EndData, State}).

post_end_per_group(_Suite, _Group, _Config, _Return, State) ->
    ?LOGGING_CALL({ok, State}).

%%% TESTCASE
pre_init_per_testcase(_Suite, _TC, InitData, State) ->
    ?LOGGING_CALL({InitData, State}).

post_init_per_testcase(_Suite, _TC, Config, _Return, State) ->
    ?LOGGING_CALL({Config, State}).

pre_end_per_testcase(_Suite, _TC, EndData, State) ->
    ?LOGGING_CALL({EndData, State}).

post_end_per_testcase(_Suite, _TC, _Config, _Return, State) ->
    ?LOGGING_CALL({ok, State}).
```

When running this, we can see that the post_xxx_per_yyy functions are executed in the wrong order. For example, for `init_per_suite/1` and `end_per_suite/1` we see:

```
----------------------------------------------------
2023-06-13 11:24:45.181
cth_a -- pre_init_per_suite
----------------------------------------------------
2023-06-13 11:24:46.182
cth_b -- pre_init_per_suite
----------------------------------------------------
2023-06-13 11:24:47.183
cth_a -- post_init_per_suite
----------------------------------------------------
2023-06-13 11:24:48.184
cth_b -- post_init_per_suite
...
----------------------------------------------------
2023-06-13 11:25:05.308
cth_b -- pre_end_per_suite
----------------------------------------------------
2023-06-13 11:25:06.308
cth_a -- pre_end_per_suite
----------------------------------------------------
2023-06-13 11:25:07.309
cth_b -- post_end_per_suite
----------------------------------------------------
2023-06-13 11:25:08.310
cth_a -- post_end_per_suite
```

**Expected behavior**
In the example above:
  - `post_init_per_suite` for `cth_b` should be called before that for `cth_a`
  - `post_end_per_suite` for `cth_a` should be called before that for `cth_b`
  - Same for all other callbacks

The current behavior seems wrong for management of contexts. Just like for the `end_per_*` functions the order is reversed wrt to the `init_per_*` calls, one would expect the order for the `post_*` functions should be reversed wrt to the `pre_*` ones.

For example, imagine you want to write a simple hook that measures some resource usage, say wall time (but could be memory, etc) during test initialization; at the moment the collected data will be wrong since, while it will account for the time spent in `pre_init_per_*` functions of all hooks with lower priority (higher numerical value), it will fail to account for the resource usage of these same hooks in their `post_init_per_*` callbacks.

**Affected versions**
Tried with OPT/25

**Comment by u3s** [comment]
I think _wrong order_ is related to original purpose of hook functionality which was intended to reimplement something that existed back then. I think it should not be changed as:
- there might be users relying on this
- their perception of _wrong order_ might be opposite

Existing behavior (A), seems to be optimal for scenario when:
- only some hooks are implement (e.g. `post_init_per_testcase` and `post_end_per_testcase`)
- the Testcase is in center of attention - therefore all hooks executed before testcase will have same order (as they're supposed to prepare resource for The Testcase) and hooks after Testcase will use reversed order

Indeed, it does not seem optimal if you make hooks focused on other test functions (B) (i.e. init_per_suite, init_per_testcase).
In other words, in (B) scenario we would like to reverse the hook order for all `post_*` hooks (at the same time ignoring if it is executed before or after Testcase).

Some years ago I worked around similar problem by splitting hook modules and adjusting their priorities separately.
This should work for your scenario as well, although being cumbersome.

Maybe we could have `ct_hooks_post_order` option for adjusting to (B) scenario.

What do you think? Can you live with the workaround?
I wonder what's the interest for (B) scenario ... years ago I assumed I was the only one interested :-D

**Comment by jcpetruzza** [comment]
To be clear, I think that having the end calls happen reversed with respect the inits is totally correct. What I find surprising is that this doesn't happen also for pre/post calls, as I'd expect post calls to be reversed wrt their pre calls. This doesn't violate test-centrism, gives you actually "callback centrism" and, more generally, makes hooks easier to compose.

I understand that changing this is likely to break some code (possibly code working this around!), so I understand if this is left as it is, at least by default.

> Maybe we could have ct_hooks_post_order option for adjusting to (B) scenario.

That would work well

> Some years ago I worked around similar problem by splitting hook modules and adjusting their priorities separately.
This should work for your scenario as well, although being cumbersome.

Yeah, I too worked this around with two instances of the hook, running with different priorities. It wouldn't have been too bad if there was a way for two hooks to share their CTHState, but well.

**Comment by u3s** [comment]
> To be clear, I think that having the end calls happen reversed with respect the inits is totally correct. What I find surprising is that this doesn't happen also for pre/post calls, as I'd expect post calls to be reversed wrt their pre calls. This doesn't violate test-centrism, gives you actually "callback centrism" and, more generally, makes hooks easier to compose.

Well, I might be missing something, but I think that with "callback centrism" enabled, code now using `post_init_per_testcase` and `post_end_per_testcase` defined on 2 hook modules with different priorities would have same execution order. This might not for work for preparing to testcase execution. 

User would need to use `pre` callback for init and `post` for end callback. This would be very uncomfortable to explain and remember.

I think combining "testcase" with "callback" centrism is not doable actually.

> I understand that changing this is likely to break some code (possibly code working this around!), so I understand if this is left as it is, at least by default.

Good point.

> > Maybe we could have ct_hooks_post_order option for adjusting to (B) scenario.
> 
> That would work well

We have some momentum with common_test now, however since there is a work around and summer period is getting closer work is expected to be down prioritized for later.

> > Some years ago I worked around similar problem by splitting hook modules and adjusting their priorities separately.
> > This should work for your scenario as well, although being cumbersome.
> 
> Yeah, I too worked this around with two instances of the hook, running with different priorities. It wouldn't have been too bad if there was a way for two hooks to share their CTHState, but well.

Yeah probably useful, but maybe unnecessarily complex?

**Comment by jcpetruzza** [comment]
> I think combining "testcase" with "callback" centrism is not doable actually.

I think I may now understand why we seem not to be agreeing here. There are actually two ways to interpret what "hook priority" means:

1. "Priority is a property of the callback". This is what CT does. By design, all callbacks in a module have the same priority (I guess for conceptual simplicity), although one could imagine an alternative one where this is not the case. It makes sense, of course, that for end callbacks the call order is reversed. 

3. "Priority is a property of the hook/module". This is what I always thought CT was doing! :slightly_smiling_face:. In this model, the hook with the highest (numerical) priority "wraps" the inits of the suite in full, the next module wraps the result of this, etc, like layers of an onion. For the ends, the wrapping happens in reversed order. What does "wrapping" mean? That for each CT callback you get something like (ignoring failure handling for simplicity):

```
init_per_suite_layer_n_plus_1(Config0) ->
  CTHState0 = magically_get_state(the_hook),
  {Config1, CTHState1} = the_hook:pre_init_per_suite(the_actual_SUITE, Config0, CTHState0),
  Config2 = init_per_suite_layer_n(Config1),
  {Config3, CTHState2} = the_hook:post_init_per_suite(the_actual_SUITE, Config2, CTHState1),
  magically_store_state(the_hook, CTHState2},
  Config3.
```

Notice that ends are reversed wrt inits because the layering order is reversed, and posts are reversed wrt pres because of the call stack.

So with "priority is a property of the callback" as mental model, my claim that the current ordering is wrong sounds, well, wrong. And if you are under the "priority is a property of the hook" model, what CT does looks absurd. Hopefully this is more clear now!

I think then the big issue is that it is currently very easy to assume "priority is a property of the hook" and build the wrong expectations about what CT will do. The [docs](https://www.erlang.org/doc/apps/common_test/ct_hooks_chapter.html) only say:
> ## CTH Execution Order
> By default, each CTH installed is executed in the order that they are installed for init calls, and then reversed for end calls. This is not always desired, so Common Test allows the user to specify a priority for each hook.

This is quite unfortunate as it fits well both models. And the "priority is a property of the hook" is not really an obscure interpretation; e.g. buck2's erlang integration comes with its own, simpler, runner of common-tests for its interactive mode and, funnily enough, hooks were implemented "wrong" there as they follow "priority is a property of the hook"! So, docs need to be much more explicit about execution order than they are now, or people will keep tripping over this.

Then there's the question of whether CT should have an optional "priority is a property of the hook" mode. I still think that it is a much preferable model overall, and CT should support it. With it, hooks become modular and compose well. E.g. this means that you can "inject" a hook deeper in the onion layer, it will get the chance to modify things set by any other hook in an outer layer, and clean-up things so that the outer hook never finds out. Hook authors can write abstractions knowing that the details won't leak and won't interact badly with other hooks. They become akin to function decorators, in a way, as you find in other languages.

**Comment by jcpetruzza** [comment]
This is a very good summary of the discussion and the diagrams are just great; e.g. looking at the second one, it is immediately clear to me that order reversal should happen, as you say, only in pre/post, but not in init/end.

Re introducing a pre_testcase/post_testcase, fwiw, that'd be a big +1 from my side: I've been simulating those two with hacks in post_init_per_testcase/pre_end_per_testcase, but it's both messy and brittle.

**Comment by u3s** [comment]
Update on next steps

> ## Next step ideas
> 1. improve existing documentation for hooks (actually it was planned for many years but down prioritized)

- reference from manual to user guide is planned to be added within #7455 
- after re-reading the user guide, I think it is quite complete :-) in a sense that it documents how the system works  (pre, post hooks wrapping around config function but when executing reversed after testcase)
- what is missing most is a diagram or two (see next step)

> 2. add mermaid diagrams to docs when that is possible in Erlang docs

- above diagrams are stored in internal common_test notes
- planned to be added to docs when mermaid and markdown is supported (work on this is started)

> 3. introduce a CT option for Configuration centric hook execution order (maybe named ct_hooks_order := [testcase(default) | configuration])

- I have some prototype already, hope to make a PR soon.

> 4. consider Testcase hook callbacks

- We will consider this, but this would be a bigger work and I don't know if we will decide to invest time in this.

**Comment by u3s** [comment]
FYI, a prototype of ct_hooks_order option available in https://github.com/erlang/otp/pull/7496

**Comment by u3s** [comment]
- #7496 is merged.
- further doc improvements are planned in OTP-27

I'm closing issue according to plan proposed above.
